sympy/integrals/heurisch.py

```diff
@@ -558,8 +558,20 @@ def find_non_syms(expr):
         if solution is None:
             return None
         else:
-            solution = [ (k.as_expr(), v.as_expr()) for k, v in solution.items() ]
-            return candidate.subs(solution).subs(list(zip(coeffs, [S.Zero]*len(coeffs))))
+            sol = {}
+            # remove coeff from k in case Floats were involved
+            # issue 8901
+            for k, v in solution.items():
+                k, v = k.as_expr(), v.as_expr()
+                if not k.is_Symbol:
+                    if k.is_Mul:
+                        c, k = k.as_coeff_Mul()
+                        v /= c
```

Comment:
I was also beginning to think that `solve_lin_sys()` should be changed. In fact, it is natural to expect that the keys were symbols.